src/System.Management.Automation/engine/Utils.cs

```diff
@@ -33,6 +34,215 @@ namespace System.Management.Automation
+        {
+            if (value < ulong.MinValue || ulong.MaxValue < value)
+            {
+                ul = 0;
+                return false;
+            }
+
+            ul = (ulong)value;
+            return true;
+        }
+
+        internal static bool TryCast(BigInteger value, out decimal dm)
+        {
+            if (value < (BigInteger)decimal.MinValue || (BigInteger)decimal.MaxValue < value)
+            {
+                dm = 0;
+                return false;
+            }
+
+            dm = (decimal)value;
+            return true;
+        }
+
+        internal static bool TryCast(BigInteger value, out double db)
+        {
+            if (value < (BigInteger)decimal.MinValue || (BigInteger)decimal.MaxValue < value)
+            {
+                db = 0;
+                return false;
+            }
+
+            db = (double)value;
+            return true;
+        }
+
+        /// <summary>
+        /// Parses a given string or ReadOnlySpan&lt;char&gt; to calculate its value as a binary number.
+        /// Assumes input has already been sanitized and only contains zeroes (0) or ones (1).
+        /// </summary>
+        /// <param name="digits">Span or string of binary digits. Assumes all digits are either 1 or 0.</param>
+        /// <param name="unsigned">
+        /// Whether to treat the number as unsigned. When false, respects established conventions
+        /// with sign bits for certain input string lengths.
+        /// </param>
+        /// <returns>Returns the value of the binary string as a BigInteger.</returns>
+        internal static BigInteger ParseBinary(ReadOnlySpan<char> digits, bool unsigned)
+        {
+            if (!unsigned)
+            {
+                if (digits[0] == '0')
+                {
+                    unsigned = true;
+                }
+                else
+                {
+                    switch (digits.Length)
+                    {
+                        // Only accept sign bits at these lengths:
+                        case 8: // byte
+                        case 16: // short
+                        case 32: // int
+                        case 64: // long
+                        case 96: // decimal
+                        case int n when n >= 128: // BigInteger
+                            break;
+                        default:
+                            // If we do not flag these as unsigned, bigint assumes a sign bit for any (8 * n) string length
+                            unsigned = true;
+                            break;
+                    }
+                }
+            }
+
+            // Calculate number of 8-bit bytes needed to hold the input,  rounded up to next whole number.
+            // This value will also be used as the indexer for our array.
+            int outputByteWalker = (digits.Length + 7) / 8;
+            byte[] outputBytes = new byte[outputByteWalker--];
```

Comment:
Surprisingly, in this instance we *did* try both, and the `Span<byte>` version was slower. Additionally, we cannot use `stackalloc` here because the span is passed to another method (the `BigInteger` constructor. I tried to do something like this in the tokenizer itself before, and got compile errors.

Reply:
@iSazonov 

Update: Re-ran the benchmarks *just* to be sure:
https://gist.github.com/vexx32/dd9532957764e74e731d1e64817b708e

It's *really* close, but I think you may be right that the span is better here -- *just*. A scant few nanoseconds difference, but improvement is improvement!!

Reply:
It is an _allocation_ optimization. Really we exclude there allocations at all, we allocate only for very-very-very huge numbers. And yes, I'd expect that stack allocation is faster then heap one.

Reply:
Right, but the compiler won't let me use stackalloc here, unless there's something I'm missing?

Reply:
Sample 
https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/FormatAndOutput/common/BaseOutputtingCommand.cs#L967-L968

Reply:
... Weird. I'm *sure* I tried something like this before... but it seems to work, thank you!